sys/include/event.h

```diff
@@ -0,0 +1,214 @@
+ * 3. events can be safely used (and actually perform best) when used within
+ *    one thread, e.g., in order to create a state-machine like process flow.
+ *    This is not (easily) possible using msg queues, as they might fill up.
+ * 4. an event can only be queued in one event queue at the same time.
+ *    Notifying many queues using only one event object is impossible.
+ *
+ * At the core, event_wait() uses thread flags to implement waiting for events
+ * to be queued. Thus event queues can be used safely and efficiently in combination
+ * with thread flags and msg queues.
+ *
+ * Examples:
+ *
+ *     // simple event handler
+ *     static void handler(event_t *event)
+ *     {
+ *        printf("triggered 0x%08x\n", (unsigned)event);
+ *     }
+ *
+ *     static event_t event = { .handler=handler };
+ *     static event_queue_t queue;
+ *
+ *     int main(void)
+ *     {
+ *         event_queue_init(&queue);
+ *         event_loop(&queue);
+ *     }
+ *
+ *     [...] event_post(&queue, &event);
+ *
+ *     // example for event extended event struct
+ *     typedef struct {
+ *         event_t super;
+ *         const char *text;
+ *     } custom_event_t;
+ *
+ *     static void custom_handler(event_t *event)
+ *     {
+ *         custom_event_t *custom_event = (custom_event_t *)event;
+ *         printf("triggered custom event with text: \"%s\"\n", custom_event->text);
+ *     }
+ *         static custom_event_t custom_event = { .super.callback=custom_handler, .text="CUSTOM EVENT" };
+ *
+ *     [...] event_post(&queue, &custom_event)
+ *
+ *
+ *
+ * @{
+ *
+ * @file
+ * @brief       Event API
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ */
+
+#ifndef EVENT_H
+#define EVENT_H
+
+#include <stdint.h>
+
+#include "irq.h"
+#include "thread_flags.h"
+#include "clist.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef THREAD_FLAG_EVENT
+/**
+ * @brief   Thread flag use to notify available events in an event queue
+ */
+#define THREAD_FLAG_EVENT   0x1
+#endif
+
+/**
+ * @brief   event_queue_t static initializer
+ */
+#define EVENT_QUEUE_INIT    { .waiter=(thread_t *)sched_active_thread }
+
+/**
+ * @brief   event structure forward declaration
+ */
+typedef struct event event_t;
+
+/**
+ * @brief   event handler type definition
+ */
+typedef void(*event_handler_t)(event_t*);
+
+/**
+ * @brief   event structure
+ */
+struct event {
+    clist_node_t list_node;     /**< event queue list entry             */
+    event_handler_t handler;    /**< pointer to event handler function  */
+};
+
+/**
+ * @brief   event queue structure
+ */
+typedef struct {
+    clist_node_t event_list;    /**< list of queued events              */
+    thread_t *waiter;           /**< thread ownning event queue         */
+} event_queue_t;
+
+/**
+ * @brief   Initialize an event queue
+ *
+ * This will set the calling thread as owner of @p queue.
+ *
+ * @param[in,out]   queue   event queue object to initialize
```

Comment:
I would argue it is just an `out` parameter. Or are there any fields required to be set before initialization.

Reply:
fixed